### PR TITLE
AudioVideoPlayer

### PR DESCRIPTION
**Pull request by tonikocjan**
An `AVPlayer` wrapper with additional functionality like:
  1. looping
  2. state (`isPlaying`)
  2. observing different events
  3. easier playback intervals manipulation,
  4. ...

### CODE REVIEW

**Review by MrNiceMrGood — APPROVED** [review]
Looks good to me

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    guard let url = URL(string: urlString) else { return nil }
+    self.init(url: url)
+  }
+}
+
+// MARK: - API
+public extension AudioVideoPlayer {
+  override func play() {
+    super.play()
+    isPlaying = true
```
Comment (borut-t) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:66]: Do we really need this flag? I would rather go with [rate](https://developer.apple.com/documentation/avfoundation/avplayer/1388846-rate?preferredLanguage=occ) instead. What do you think?
```swift
var isPlaying: Bool {
  rate != 0
}
```

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    currentItem?.asset.duration.seconds ?? 0
+  }
+}
+
+// MARK: - Private Methods
+private extension AudioVideoPlayer {
+  func setupTimeObserver() {
+    guard timeObserver == nil else { return }
+
+    let timeInterval = CMTimeMake(value: 1, timescale: 10)
```
Comment (borut-t) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:120]: So the observer will trigger every 1/10th of a second. What if we need it to be triggered more/less often? Should we make this configurable?

**Review by borut-t — COMMENTED** [review]
I like it. Just added two comments to think about 🤔

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift (hunk 1 above)
Comment (tonikocjan) [reply Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:66]: This looks very promising, yes. 👍  Will test it out if it works as expected.

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift (hunk 2 above)
Comment (tonikocjan) [reply Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:120]: We should, of course.

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    setupTimeObserver()
+  }
+
+  var duration: Double {
+    currentItem?.asset.duration.seconds ?? 0
+  }
+  
+  var isPlaying: Bool {
+    rate != 0.0
+  }
```
Comment (gentique) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:127]: It seems that it's possible for the player to stall due to buffer issues, etc. In that case the rate seems to not go to 0. We should also check if there's any error.

```suggestion
  var isPlaying: Bool {
    rate != 0.0 && error == nil
  }
```

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    playbackInterval = (startAt, endAt)
+  }
+
+  func setPlaybackPosition(
+    startAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
+    removeTimeObserver()
+    seek(to: CMTime(seconds: startAt, preferredTimescale: timescale)) { _ in
+      self.setupTimeObserver()
```
Comment (gentique) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:109]: Is it possible to have a retain cycle here?

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift (hunk 3 above)
Comment (tonikocjan) [reply Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift]: Used `status` instead, but very good point.

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift (hunk 4 above)
Comment (tonikocjan) [reply Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:109]: The completion handler is called when `seek` either succeeds or is interrupted, so I think not.

**Review by gentique — APPROVED** [review]

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    removeTimeObserver()
+    setupTimeObserver()
+  }
+
+  /// Set `playbackInterval` and force player to start playing from the `startAt` timestamp.
+  func setPlaybackInterval(
+    startAt: Double,
+    endAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
```
Comment (borut-t) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:97]: Please rename to `play` and update the comment to
```swift
/// Starts and ends playing at `startAt` and `endAt` respectively.
```

File: Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    guard startAt < endAt else { fatalError("`startAt` should be less than `endAt`") }
+    setPlaybackPosition(startAt: startAt)
+    playbackInterval = (startAt, endAt)
+  }
+
+  func setPlaybackPosition(
+    startAt: Double,
+    timescale: CMTimeScale = AudioVideoPlayer.defaultTimescale
+  ) {
```
Comment (borut-t) [inline Sources/Utilities/AudioVideoPlayer/AudioVideoPlayer.swift:106]: Please rename to `play` and add a comment
```swift
/// Starts playing at `startAt`.
```

**Review by borut-t — APPROVED** [review]
Two notes and we are ready.

**Comment by tonikocjan** [comment]
I'm closing this one for the time being. Let's maybe consider it again in the future.

**Pull request closed without merging**